providers/google.go

```diff
@@ -185,51 +189,46 @@ func (g *GoogleOp) requestTokens(ctx context.Context, cicHash string) (*simpleoi
 			return
 		}
 
-		chTokens <- retTokens
-
-		// If defined the OIDC client hands over control of the HTTP server session to the OpenPubkey client.
-		// Useful for redirecting the user's browser window that just finished OIDC Auth flow to the
-		// MFA Cosigner Auth URI.
+		// If defined the OIDC client hands over control of the HTTP server
+		// session to the OpenPubkey client. Useful for redirecting the user's
+		// browser window that just finished OIDC Auth flow to the MFA Cosigner
+		// Auth URI.
 		if g.httpSessionHook != nil {
 			g.httpSessionHook(w, r)
-			defer shutdownServer() // If no http session hook is set, we do server shutdown in RequestTokens
 		} else {
 			if _, err := w.Write([]byte("You may now close this window")); err != nil {
 				logrus.Error(err)
 			}
 		}
+
+		chTokens <- retTokens
 	}
 
 	callbackPath := redirectURI.Path
-	http.Handle(callbackPath, rp.CodeExchangeHandler(marshalToken, relyingParty))
+	mux.Handle(callbackPath, rp.CodeExchangeHandler(marshalToken, relyingParty))
 
 	go func() {
 		err := g.server.Serve(ln)
 		if err != nil && err != http.ErrServerClosed {
 			logrus.Error(err)
 		}
 	}()
+	defer func() {
```

Comment:
I don't think this approach works with the cosigner.
1. requestTokens exits, defer called, server shuts down,
2. cosigner flow starts and calls HookHTTPSession
3. HookHTTPSession attempts to do the redirect using the server but the server is shutdown.  Error!

I think this breaks the cosigner flow

Reply:
Interesting. For some reason, I thought everything was fine because previous code looks like this:

https://github.com/openpubkey/openpubkey/blob/eba749e7d12b83798f2ab29ed684ba3bafed5e6f/providers/google.go#L190-L200

the `hook` is called and then once `hook` func returns `serverShutdown()` is called immediately (this is in `callbackPath` handler).

With the changes in this PR, I flipped the order so that the hook happens first and then we write to `chTokens` channel

https://github.com/ymarcus93/openpubkey/blob/878cab32fd7555dde1a0b55dd8fa16a65742bbae/providers/google.go#L192-L204

Once `chTokens` is written to `requestToken` exits and `defer` is called and server shuts down. But by that point shouldn't the `hookHttpSession` method already have been executed?

Reply:
I guess my question is when does the cosigner flow start? When `httpSessionHook` is called or after `requestToken` exits?

Reply:
Cosigner flow:
1. Cosigner sets httpSessionHook hook,
2. oidcAuth called
  3. Oidc Auth calls OP.requestTokens
  4. OP.requestTokens finishes
  5. oidcAuth finishes
6. Cosigner flow starts
7. Cosigner flow uses hook to redirect user's open webpage to cosigner init auth
8. After init auth completes, OP http server is shutdown
9. Cosigner flow finishes

Reply:
The line
```go
 if g.httpSessionHook != nil { 
 	g.httpSessionHook(w, r) 
 	defer shutdownServer() // If no http session hook is set, we do server shutdown in RequestTokens 
 ```
 
 Is happening inside marshalToken after `chTokens <- retTokens` and `chTokens <- retTokens` causes requestTokens to return.

Reply:
> I guess my question is when does the cosigner flow start? When httpSessionHook is called or after requestToken exits?

The cosigner code starts when requestToken exits, but the flow is kicked off by using the channel in httpSessionHook to redirect the user's web browser

1. Client.Auth sets the hook

https://github.com/openpubkey/openpubkey/blob/eba749e7d12b83798f2ab29ed684ba3bafed5e6f/client/client.go#L188C1-L200C4

2. Client.Auth then calls requestTokens

3. requestTokens does the OIDC flow and returns, but marshalTokens is still blocking on the hook

4. After requestTokens returns, the Cosigner flow starts,

5. The Cosigner flow uses the hook to redirect the http server and unblocks marshalTokens

I dislike how complex this is, but I haven't found a better way of doing it.